### PR TITLE
Refactor FilterPath parse function (#12067)

### LINKED ISSUE #12067: [Refactor] Replace recursive FilterPath.parse with iterative approach
### Describe the bug

This issue is related to a recent pull request: https://github.com/opensearch-project/OpenSearch/pull/11060

There is another instance of recursion used in the [FilterPath.parse](https://github.com/opensearch-project/OpenSearch/blob/main/libs/core/src/main/java/org/opensearch/core/xcontent/filtering/FilterPath.java#L112-L128) method which appears inefficient and unnecessary. It appears to be ignoring repeat `.` until the last is located and could more optimally skip until the last `.` in the sequence instead of recursing.  

### Additional Context

In https://github.com/opensearch-project/OpenSearch/pull/11060, there was a recursive method for evaluating core's own [Regex](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/regex/Regex.java) class. The method [simpleMatchWithNormalizedString](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/regex/Regex.java#L131-L153) was refactored from a recursive approach to an iterative approach for more predictable space and runtime complexity.

- [ ] Refactor FilterPath.parse to use iterative method
- [ ] Update FilterPathTests to cover more scenarios

### Related component

Search

### To Reproduce

Should maintain same behavior as now

### Expected behavior

Should maintain same behavior as now

### Additional Details

_No response_

### PR DESCRIPTION
**Pull request by robinf95**
Signed-off-by: Robin Friedmann <robinfriedmann.rf@gmail.com>

### Description
Refactoring [FilterPath.parse](https://github.com/opensearch-project/OpenSearch/blob/main/libs/core/src/main/java/org/opensearch/core/xcontent/filtering/FilterPath.java#L112-L128) by using an iterative approach instead of recursion.

### Related Issues
Resolves #12067 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: libs/core/src/main/java/org/opensearch/core/xcontent/filtering/FilterPath.java [hunk 1]
```diff
@@ -99,32 +99,26 @@ public static FilterPath[] compile(Set<String> filters) {
...
         List<FilterPath> paths = new ArrayList<>();
         for (String filter : filters) {
-            if (filter != null) {
+            if (filter != null && !filter.isEmpty()) {
                 filter = filter.trim();
                 if (filter.length() > 0) {
-                    paths.add(parse(filter, filter));
+                    paths.add(parse(filter));
                 }
             }
```
Comment (deshsidd) [inline libs/core/src/main/java/org/opensearch/core/xcontent/filtering/FilterPath.java:107]: Warning log in the else case required?

File: libs/core/src/main/java/org/opensearch/core/xcontent/filtering/FilterPath.java [hunk 2]
```diff
@@ -99,32 +99,26 @@ public static FilterPath[] compile(Set<String> filters) {
...
         }
         return paths.toArray(new FilterPath[0]);
     }
 
-    private static FilterPath parse(final String filter, final String segment) {
-        int end = segment.length();
+    private static FilterPath parse(final String filter) {
+        String[] segments = filter.split("(?<!\\\\)\\.");
```
Comment (deshsidd) [inline libs/core/src/main/java/org/opensearch/core/xcontent/filtering/FilterPath.java:113]: nit: Could add comments her for better readability? 
Eg: // Split the filter into segments using a regex that avoids splitting escaped dots.

File: libs/core/src/main/java/org/opensearch/core/xcontent/filtering/FilterPath.java [hunk 3]
```diff
@@ -99,32 +99,26 @@ public static FilterPath[] compile(Set<String> filters) {
...
-            if (c == '.') {
-                String current = segment.substring(0, i).replaceAll("\\\\.", ".");
-                return new FilterPath(filter, current, parse(filter, segment.substring(i + 1)));
-            }
-            ++i;
-            if ((c == '\\') && (i < end) && (segment.charAt(i) == '.')) {
-                ++i;
-            }
+        for (int i = segments.length - 1; i >= 0; i--) {
+            String segment = segments[i].replaceAll("\\\\.", ".");
```
Comment (deshsidd) [inline libs/core/src/main/java/org/opensearch/core/xcontent/filtering/FilterPath.java:117]: // Replace escaped dots with actual dots in the current segment.

**Comment by robinf95** [comment]
> Do we have tests covering these already?

There are already a bunch of tests [FilterPathTests](https://github.com/opensearch-project/OpenSearch/blob/main/libs/core/src/main/java/org/opensearch/core/xcontent/filtering/FilterPathTests.java)

I just added two that were missing.

**Review by scrawfor99 — APPROVED** [review]

**Comment by getsaurabh02** [comment]
@deshsidd will you be interested in taking this forward?

**Comment by deshsidd** [comment]
> @deshsidd will you be interested in taking this forward?

Let me take care of this one

**Comment by sandeshkr419** [comment]
@deshsidd Are you still on it?

File: libs/core/src/test/java/org/opensearch/core/xcontent/filtering/FilterPathTests.java [hunk 4]
```diff
@@ -369,4 +370,20 @@ public void testMultipleFilterPaths() {
         assertThat(filterPath.getSegment(), is(emptyString()));
         assertSame(filterPath, FilterPath.EMPTY);
     }
+
+    public void testCompileWithEmptyString() {
```
Comment (gaobinlong) [inline libs/core/src/test/java/org/opensearch/core/xcontent/filtering/FilterPathTests.java:374]: Seems this newly added test case is duplicated with the existing test case: https://github.com/opensearch-project/OpenSearch/blob/14f1c43c108f378b13d109ade364216c082fb858/libs/core/src/test/java/org/opensearch/core/xcontent/filtering/FilterPathTests.java#L122.

File: libs/core/src/test/java/org/opensearch/core/xcontent/filtering/FilterPathTests.java [hunk 5]
```diff
@@ -369,4 +370,20 @@ public void testMultipleFilterPaths() {
...
+        Set<String> filters = new HashSet<>();
+        filters.add("");
+        FilterPath[] filterPaths = FilterPath.compile(filters);
+        assertNotNull(filterPaths);
+        assertEquals(0, filterPaths.length);
+    }
+
+    public void testCompileWithNull() {
```
Comment (gaobinlong) [inline libs/core/src/test/java/org/opensearch/core/xcontent/filtering/FilterPathTests.java:382]: Same to above, duplicated.

**Comment by sandeshkr419** [comment]
Closing in lieu of https://github.com/opensearch-project/OpenSearch/pull/14200

**Pull request closed without merging**